### ISSUE TITLE
[SomeChanges] Make stronger versions of Demonic Pact overwrite weaker ones.

### ISSUE THREAD

**Issue by Zephancode** [Zephancode/wotlk-warlock#28]
## 1. Is your feature request related to a problem? Please describe.
One of the core reasons Demonology Warlocks are brought to raids is [Demonic Pact](https://www.wowhead.com/wotlk/spell=48090/demonic-pact). The buff provides the whole raid with 10% of the Warlock's spell power. This buff gets applied whenever the warlock's pet lands a critical hit and stays up for the next 45 seconds. It also incurs a 20-second internal cooldown, as such if the pet crits again during this time Demonic Pact does not get refreshed. After this 20 seconds duration, weaker versions of Demonic Pact (e.g. trigger when the warlock had less spell power) can overwrite an already stronger version that is present (from the same player).

This makes it so that if the Demonology Warlock wants to keep the stronger version (for example when a trinket, blood fury, etc.. is up) active after the initial 20 seconds (after the ICD is up) he is forced to call back his pet and put him on passive until the buff is about to run out and send him back. Leaving the pet to AFK for 25 seconds out of the 45 seconds the buff is up, or in other words over half the time. 

This behavior is observed in the [current beta](https://github.com/Zephancode/wotlk-warlock/issues/26) as can be seen here and was the case back in [retail wotlk as can be seen mentioned in this elitest jerk post](https://web.archive.org/web/20100807225539/http://elitistjerks.com/f80/t100099-demonology_releasing_demon_you/#Demonic_Pact).

What this facilitates is known as "Pet twisting" where a Demonology warlock sacrifices personal DPS by recalling their pet once Demonic Pact is up, in order not to overwrite it with a weaker version. Followed by sending the pet back in when Demonic Pact is about to run out.

When multiple Demonology Warlocks are present stronger versions can overwrite weaker versions of the other warlock, BUT weaker versions can not overwrite stronger ones from the other warlock.

### 1.1 Describe the solution you'd like
In an ideal world the same methodology that applies to having multiple warlocks would be present for a single warlock as well. This would mean that there is no need for "Pet twisting". In other words: Stronger versions take precedent over weaker ones. This would promote actually using your pet to deal damage instead of sidelining your demon for over half the fight.

### 1.2 Additional context
A lot of warlocks want to try out the revamped Demonology talent tree but are put off by the idea of having to "Pet Twist".

**Comment by NoCement** [comment]
I agree that this mechanic sadly pushes Demonology warlocks into very obnoxious and boring gameplay and discourages people from playing Demonology in general (based on feedback from other players). Majority of the Demonology talent tree directly or inderectly buffs your pet only to be forced by this design to not have pet attack half of the time in raid environment.

I would like to think this gameplay was really not intended when designing Demonic Pact.

Apart from suggestions above, making DP dynamic with your current Spell Power value and not snapshot Spell Power when activated for it's whole duration would solve this issue.

**Comment by F-Joubert** [comment]
I think this is a fun part of the skill expression and it's not like your raid needs more than one buffbot demo lock anyway so you're unlikely to be forced into this playstyle if you don't want to be. There are other versions of demo lock. Or be a FOTM afflock, just don't be put off by corruption snapshotting.

**Comment by nodddle** [comment]
ideally only stronger DP buffs refresh for the first 20-30 seconds of the buff, then after that weaker versions are allowed refresh so the buff doesn't have downtime, you get to keep your pet in, and you get to provide the strongest DP buff to your raid for the longest period of time.

you won't stand out by pet twisting, it will be the norm in hardcore guilds and you will stand out for not pet twisting.

 in what way is the warlock class identity one that's unselfish and designed to sacrifice personal dps to support the raid? I can't think of a single dps spec in tbc that supports less than warlock, maybe mage does idk. I see warlocks as a class that seeks nothing but chaos and the destruction of their enemies through the means of dark arts, not a noble, self-sacrificial, and virtuous class that is always willing to take one for the team... have you ever tried to get other warlocks to use group necks?

**Comment by ragingpastry** [comment]
It would be kinda neat if blizzard made Demonic Pact a pet ability that could be cast, instead of a proc on a pet crit. This would allow people to keep the "skill" aspect while still allowing the pet to stay in all the time. I doubt Blizzard would be able to implement this as its a fairly big change, but in my mind it is the best solution.

**Comment by esuvii** [comment]
You could imagine Demonic Pact being added to the pet bar, and you could right click it to toggle it disabled/enabled - visually indicated by the auto-cast glow other pet abilities have.

This would keep the nuanced gameplay of managing your procs and Demonic Pact timing, by toggling it, but would mean you don't need to make your pet passive to prevent early applications of Demonic Pact.

Instead of managing pet attack/passive you could use a macro like:
```
/petautocastoff Demonic Pact
/petautocaston [mod:shift] Demonic Pact
```
Or even simpler just `/cast Demonic Pact` could toggle it.

**Comment by NoCement** [comment]
I have raised https://eu.forums.blizzard.com/en/wow/t/somechanges-demonology-warlock-on-wrath-classic-demonic-pact/367292 to bring more attention to this. Any feedback to the post is appreciated.

https://us.forums.blizzard.com/en/wow/t/somechanges-demonology-warlock-on-wrath-classic-demonic-pact/1275215 for us users

**Comment by Erodraeda** [comment]
I think one good solution as well would be making the ICD and Duration be closer together, such as 40 second ICD and 45 second duration. That would make your pet be able to attack on a higher uptime, maybe even 100%, and leave you 5 seconds to reapply without risking a downtime.

And the "kill command" like idea is also really good, where for some time after your pet crits you can use the Demonic Pact and apply the buff, giving the player more control

**Comment by nodddle** [comment]
5s to reapply isn't nearly enough time. with imp fireball is 2-2.5s cast time base, if you play it right you can get 3 chances for a fireball to crit before DP falls off. people will use felguard most of the time probably which attacks faster, but even with that it's not nearly enough for reliable 100% DP uptime, it's like 4 or 5 attacks. but it's not that bad of an idea, leaving you something like 10-15s would help fix the problem but it's still not ideal.
